_posts/2017-06-15-grpc-load-balancing.md

```diff
@@ -0,0 +1,190 @@
+  </tr>
+</table>
+
+
+### Proxy Load Balancer options - L3/L4 (Transport) vs L7 (Application) level
+
+Proxy load balancing can be L3/L4 (transport level) or L7 (application level). In transport level load balancing, the server terminates the TCP connection and opens another connection to the backend of choice. The application data (HTTP/2 and gRPC frames) are simply copied between the client connection to the backend connection. L3/L4 LB by design does very little processing, adds less latency compared with  L7 LB and consumes less resources (cheaper). 
+
+In L7 (application level) load balancing, the LB terminates and parses the HTTP/2 protocol. It creates a new HTTP/2 connection to the backend of choice. It then forwards the HTTP/2 streams received from the client to the backend(s) of choice. With HTTP/2 LB can distribute the streams from one client among multiple backends. LB can inspect the request and assign backend based on request contents. As an example, a session cookie sent as part of HTTP header can be used to associate with a specific backend, so all requests for that session are served by the same backend.
+
+#### L3/L4 (Transport) vs L7 (Application)
+
+<table>
+  <tr>
+    <td>
+    Use case
+    </td>
+    <td>
+    Recommendation
+    </td>
+  </tr>
+  <tr>
+    <td>RPC load varies a lot among connections</td>
+    <td>Use Application level LB</td>
+  </tr>
+  <tr>
+    <td>Storage or compute affinity is important</td>
+    <td>Use Application level LB and use cookies or similar for routing requests to correct backend</td>
+  </tr>
+  <tr>
+    <td>Minimizing resource utilization in proxy is more important than features</td>
+    <td>Use L3/L4 LB</td>
+  </tr>
+  <tr>
+    <td>Latency is paramount</td>
+    <td>Use L3/L4 LB</td>
+  </tr>
+</table>
+
+
+### Client side LB options - Thick client vs Look-aside
+
+*Note: look-aside is also known as one-arm in some literature.*
+
+#### Thick client
+
+A thick client approach means the load balancing smarts are implemented in the client. The client is responsible for keeping track of available servers, their workload, and the algorithms used for choosing servers. The client typically integrates libraries that communicate with other infrastructures - Service Discovery, Name Resolution, Quota management, etc.
+
+#### Lookaside Load Balancing
+
+A lookaside load balancer is another approach where the smarts are implemented in a different LB server. Clients query the lookaside LB and the LB responds with best server(s) to use. The heavy lifting of keeping server state and implementation of LB algorithm is consolidated in the lookaside LB. Note that client might choose to implement simple algorithms on top of the sophisticated ones implemented in the LB. GRPC defines a protocol for communication between client and LB. See this [doc](https://github.com/grpc/grpc/blob/master/doc/load-balancing.md) for details.
```

Comment:
With lookaside load balancing, the load balancing smarts are implemented in a special LB server.
...
Note that a client might choose to implement simple algorithms on top of the sophisticated ones implemented in the LB. gRPC defines a protocol for communication between the client and the LB using this model. See [Load Balancing in gRPC](https://github.com/grpc/grpc/blob/master/doc/load-balancing.md) for details.